ARROW-8151: [Dataset][Benchmarking] benchmark S3File performance

lidavidm opened the pull request:
This does some basic refactoring of the S3 filesystem test to share some code, but I'd appreciate any feedback on how to better organize things.

This currently has 4 sets of benchmarks:
- A test of reading an entire file of various sizes from S3
- A test of reading sequential small chunks of those files
- A test of using coalescing to read those sequential small chunks
- A test of reading a Parquet file

lidavidm commented:
I ran the benchmark both with and without connection latency, introduced via [toxiproxy](https://github.com/Shopify/toxiproxy/). Latency added was 20ms +/- 2ms jitter (what I saw to Amazon S3 from my laptop).

Toxiproxy commands:

```
./toxiproxy-cli-linux-amd64 create minio --listen 127.0.0.1:50052 --upstream 127.0.0.1:50051
./toxiproxy-cli-linux-amd64 toxic add minio -n latency -t latency -a latency=20 -a jitter=2
```

Reading the entire file in one shot (`ReadAll`):

```
--------------------------------------------------------------------------------------------
Benchmark                                                     Time           CPU Iterations
--------------------------------------------------------------------------------------------
MinioFixture/ReadAll1Mib/min_time:15.000/real_time      4524154 ns    2257122 ns       4682   221.036MB/s    221.036 items/s
MinioFixture/ReadAll100Mib/min_time:15.000/real_time  191434898 ns  181662703 ns        109   522.371MB/s    5.22371 items/s
MinioFixture/ReadAll500Mib/min_time:15.000/real_time  885562963 ns  860429294 ns         24   564.613MB/s    1.12923 items/s

With Latency
--------------------------------------------------------------------------------------------
MinioFixture/ReadAll1Mib/min_time:15.000/real_time     46570687 ns    2750293 ns        447   21.4727MB/s    21.4727 items/s
MinioFixture/ReadAll100Mib/min_time:15.000/real_time  276031415 ns  226440164 ns         75   362.278MB/s    3.62278 items/s
MinioFixture/ReadAll500Mib/min_time:15.000/real_time 1185247456 ns 1114922703 ns         18   421.853MB/s   0.843706 items/s
```

Reading the file in chunks:

```
------------------------------------------------------------------------------------------------
Benchmark                                                         Time           CPU Iterations
------------------------------------------------------------------------------------------------
MinioFixture/ReadChunked100Mib/min_time:15.000/real_time  196150922 ns  152319302 ns        106   509.812MB/s    5.09812 items/s
MinioFixture/ReadChunked500Mib/min_time:15.000/real_time  942929100 ns  762863911 ns         22   530.263MB/s    1.06053 items/s

With latency
------------------------------------------------------------------------------------------------
MinioFixture/ReadChunked100Mib/min_time:15.000/real_time  701457618 ns  201526399 ns         29    142.56MB/s     1.4256 items/s
MinioFixture/ReadChunked500Mib/min_time:15.000/real_time 3397605245 ns  986615721 ns          6   147.162MB/s   0.294325 items/s
```

Reading the file in chunks, with coalescing:

```
--------------------------------------------------------------------------------------------------
Benchmark                                                           Time           CPU Iterations
--------------------------------------------------------------------------------------------------
MinioFixture/ReadCoalesced100Mib/min_time:15.000/real_time  175538760 ns   17630239 ns        113   569.675MB/s    5.69675 items/s
MinioFixture/ReadCoalesced500Mib/min_time:15.000/real_time  746773493 ns   83969652 ns         28   669.547MB/s    1.33909 items/s

With Latency
--------------------------------------------------------------------------------------------------
MinioFixture/ReadCoalesced100Mib/min_time:15.000/real_time  249571387 ns   18055469 ns         83   400.687MB/s    4.00687 items/s
MinioFixture/ReadCoalesced500Mib/min_time:15.000/real_time 1073518943 ns   87166607 ns         20   465.758MB/s   0.931516 items/s
```

Parquet:

```
----------------------------------------------------------------------------------------------
Benchmark                                                       Time           CPU Iterations
----------------------------------------------------------------------------------------------
MinioFixture/ReadParquet250K/min_time:15.000/real_time 1339233367 ns 1156336875 ns         16   162.304MB/s   0.746696 items/s

With throttling
----------------------------------------------------------------------------------------------
MinioFixture/ReadParquet250K/min_time:15.000/real_time 3816421753 ns 1417826696 ns          5   56.9546MB/s   0.262026 items/s
```

Read coalescing helps here, and could potentially do more - I didn't tune the parameters.

pitrou commented on cpp/src/arrow/filesystem/s3fs_benchmark.cc:
```diff
@@ -0,0 +1,317 @@
+      std::shared_ptr<Array> values;
+      std::vector<double> c_values;
+      arrow::random_real(num_rows, static_cast<uint32_t>(i), -1.e10, 1e10, &c_values);
+      arrow::NumericBuilder<DoubleType> builder(std::make_shared<DoubleType>(),
```
Can use `arrow/testing/random.h` instead.

lidavidm replied:
Good call, thanks.

pitrou commented on cpp/src/arrow/filesystem/s3_test_util.h:
```diff
@@ -0,0 +1,158 @@
+#include "arrow/filesystem/s3fs.h"
+#include "arrow/status.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/util.h"
```
Is this include actually used here ?

lidavidm replied:
It's not, I've removed it.

pitrou commented on cpp/src/arrow/filesystem/CMakeLists.txt:
```diff
@@ -36,6 +36,15 @@ if(ARROW_S3)
+
+  if(ARROW_BUILD_BENCHMARKS AND ARROW_PARQUET)
+    add_arrow_benchmark(s3fs_benchmark PREFIX "arrow-filesystem")
+    if(ARG_STATIC_LINK_LIBS)
```
I don't think `ARG_STATIC_LINK_LIBS` exists at the top level?

lidavidm replied:
I've changed this to look at `ARROW_TEST_LINKAGE` instead.

pitrou commented on cpp/src/arrow/filesystem/s3fs_benchmark.cc:
```diff
@@ -0,0 +1,317 @@
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    int64_t size = 0;
+    // N.B. this is an extra call to S3 and influences the
```
Do you mean the GetSize call? It doesn't call into S3.

lidavidm replied:
Opening the file makes a HEAD request, but I've removed the comment since it is irrelevant.

pitrou commented on cpp/src/arrow/filesystem/s3fs_benchmark.cc:
```diff
@@ -0,0 +1,317 @@
+BENCHMARK_DEFINE_F(MinioFixture, ReadAll1Mib)(benchmark::State& st) {
+  NaiveRead(st, fs_.get(), "bucket/bytes_1mib");
+}
+BENCHMARK_REGISTER_F(MinioFixture, ReadAll1Mib)->MinTime(15)->UseRealTime();
```
Are 15 seconds necessary for benchmark stability? Usually we don't specify min time in benchmark source and instead let the user pass `--benchmark_min_time` on the command line.

lidavidm replied:
In general yes, we need a much higher min time in order to run more than 1 iteration of these benchmarks, but I've removed it in favor of the CLI switch.

lidavidm commented:
Thanks for the review. I've addressed the feedback and also added a few flags so that you can choose the bucket name and skip file creation. I will run the tests against S3 and report on the mailing list discussion once I get my AWS account activated.

pitrou commented on cpp/src/arrow/filesystem/s3_test_util.h:
```diff
@@ -0,0 +1,157 @@
+#include <string>
+#include <utility>
+
+// boost/process/detail/windows/handle_workaround.hpp doesn't work
```
Why this comment, by the way?

lidavidm replied:
It originates in f8868686bb; I copied some of the includes and associated comments out of `s3fs_test.cc` and put them here when I refactored out the minio utilities.

pitrou commented on cpp/src/arrow/filesystem/s3fs_benchmark.cc:
```diff
@@ -0,0 +1,344 @@
+
+    const char* skip_str = std::getenv(kEnvSkipSetup);
+    const std::string skip = skip_str ? std::string(skip_str) : "";
+    if (skip == "true") {
```
Shouldn't it be skipped if the variable has any non-empty value?

lidavidm replied:
Sounds good to me.

pitrou commented on cpp/src/arrow/filesystem/s3fs_benchmark.cc:
```diff
@@ -0,0 +1,344 @@
+    std::shared_ptr<io::OutputStream> sink;
+    ARROW_ASSIGN_OR_RAISE(sink, fs_->OpenOutputStream(path));
+    auto writer_properties =
+        parquet::WriterProperties::Builder().write_batch_size(128)->build();
```
Does `write_batch_size` influence the row group size or something? 128 rows at once sounds tiny.

lidavidm replied:
I copied some of this out of the unit tests. I've changed this to just use defaults for everything.

pitrou commented on cpp/src/arrow/filesystem/s3fs_benchmark.cc:
```diff
@@ -0,0 +1,344 @@
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
```
`std::cerr` is used in other parts. Is there a reason for the discrepancy?

lidavidm replied:
An oversight on my part; I've made these all cerr.

lidavidm commented:
@pitrou thanks for the reviews, any further comments?

pitrou commented:
Thank you @lidavidm . Will merge.
